### DISCUSSION TITLE
Custom Functions: passing with env map to `Run` or creating in `Compile` with `expr.Function`

### DISCUSSION THREAD

**Discussion by aldas** [antonmedv/expr#542] [Q&A] [answered]
1) Which is the recommended way to deal with custom functions.  
a) passing with env map to `Run`
b) creating in `Compile` with `expr.Function`

2) Is there a performance difference?

**Comment by antonmedv** [comment, marked as answer]
> Which is the recommended way to deal with custom functions.

Expr has three ways to define function: via map, via methods on env struct, via `expr.Function`.

> Is there a performance difference?

The fastest way is via `expr.Function`.

**Reply by aldas** [reply]
> Expr has three ways to define function: via map, via methods on env struct, via expr.Function.

In which situation you prefer use which one?  Maybe some example of some situation where one is more suitable than the others?
Is there functional difference?

**Reply by antonmedv** [reply]
It depends on your data structures. For example, Order{} structure.
